build-logic/asciidoc-extensions/src/main/java/org/spockframework/plugins/asciidoctor/IncludedSourceLinker.java

```diff
@@ -0,0 +1,257 @@
+package org.spockframework.plugins.asciidoctor;
+
+import org.asciidoctor.ast.*;
+import org.asciidoctor.extension.IncludeProcessor;
+import org.asciidoctor.extension.PreprocessorReader;
+import org.asciidoctor.extension.Treeprocessor;
+import org.asciidoctor.log.LogRecord;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.lang.String.join;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.emptyMap;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.*;
+import static org.asciidoctor.extension.Contexts.LISTING;
+import static org.asciidoctor.extension.Contexts.LITERAL;
+import static org.asciidoctor.log.Severity.WARN;
+
+public class IncludedSourceLinker {
+  private static final String INCLUDE_SOURCE_MARKER_PADDING = "                                                                                                    ";
+  private final String includeSourceMarker = UUID.randomUUID().toString();
+  private final int includeSourceMarkerLength = includeSourceMarker.length();
+
+  public IncludeProcessor getIncludeProcessor() {
+    return new IncludeProcessor() {
+      private static final Pattern TAG_DELIMITER = Pattern.compile("[,;]");
+      private static final Pattern TAG_LINE_PATTERN = Pattern.compile("\\b(?<boundary>tag|end)::(?<tag>\\S+?)\\[](?: |$)");
+
+      @Override
+      public boolean handles(String target) {
+        return target.startsWith("../");
+      }
+
+      @Override
+      public void process(Document document, PreprocessorReader reader, String target, Map<String, Object> attributes) {
```

Comment:
This method is quite long. Can you break it up in descriptively named methods? Or, at least add section comments about what each section does.